node-js-basics-101/node_modules/ws/lib/Receiver.js

```diff
@@ -0,0 +1,555 @@
+        return;
+      }
+
+      if (this.pushFragment(buf)) this.dataMessage();
+      this.startLoop();
+    });
+  }
+
+  /**
+   * Handles a data message.
+   *
+   * @private
+   */
+  dataMessage () {
+    if (this.fin) {
+      const messageLength = this.messageLength;
+      const fragments = this.fragments;
+
+      this.totalPayloadLength = 0;
+      this.messageLength = 0;
+      this.fragmented = 0;
+      this.fragments = [];
+
+      if (this.opcode === 2) {
+        var data;
+
+        if (this.binaryType === 'nodebuffer') {
+          data = toBuffer(fragments, messageLength);
+        } else if (this.binaryType === 'arraybuffer') {
+          data = toArrayBuffer(toBuffer(fragments, messageLength));
+        } else {
+          data = fragments;
+        }
+
+        this.onmessage(data, { masked: this.masked, binary: true });
+      } else {
+        const buf = toBuffer(fragments, messageLength);
+
+        if (!isValidUTF8(buf)) {
+          this.error(new Error('invalid utf8 sequence'), 1007);
+          return;
+        }
+
+        this.onmessage(buf.toString(), { masked: this.masked });
+      }
+    }
+
+    this.state = GET_INFO;
+  }
+
+  /**
+   * Handles a control message.
+   *
+   * @param {Buffer} data Data to handle
+   * @private
+   */
+  controlMessage (data) {
+    if (this.opcode === 0x08) {
+      if (data.length === 0) {
+        this.onclose(1000, '', { masked: this.masked });
+        this.loop = false;
+        this.cleanup(this.cleanupCallback);
+      } else if (data.length === 1) {
+        this.error(new Error('invalid payload length'), 1002);
+      } else {
+        const code = data.readUInt16BE(0, true);
+
+        if (!ErrorCodes.isValidErrorCode(code)) {
+          this.error(new Error(`invalid status code: ${code}`), 1002);
+          return;
+        }
+
+        const buf = data.slice(2);
+
+        if (!isValidUTF8(buf)) {
+          this.error(new Error('invalid utf8 sequence'), 1007);
+          return;
+        }
+
+        this.onclose(code, buf.toString(), { masked: this.masked });
+        this.loop = false;
+        this.cleanup(this.cleanupCallback);
+      }
+
+      return;
+    }
+
+    const flags = { masked: this.masked, binary: true };
+
+    if (this.opcode === 0x09) this.onping(data, flags);
+    else this.onpong(data, flags);
+
+    this.state = GET_INFO;
+  }
+
+  /**
+   * Handles an error.
+   *
+   * @param {Error} err The error
+   * @param {Number} code Close code
+   * @private
+   */
+  error (err, code) {
+    this.onerror(err, code);
+    this.hadError = true;
+    this.loop = false;
+    this.cleanup(this.cleanupCallback);
+  }
+
+  /**
+   * Checks payload size, disconnects socket when it exceeds `maxPayload`.
+   *
+   * @param {Number} length Payload length
+   * @private
+   */
+  maxPayloadExceeded (length) {
+    if (length === 0 || this.maxPayload < 1) return false;
+
+    const fullLength = this.totalPayloadLength + length;
```

Comment:
__[jshint]__
* 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

<!-- Comment by [SideCI](https://sideci.com) -->